src/core/DataFile.cpp

```diff
@@ -1542,3 +1514,29 @@ void findIds(const QDomElement& elem, QList<jo_id_t>& idList)
 		child = child.nextSiblingElement();
 	}
 }
+
+const unsigned int DataFile::legacyFileVersion()
+{
+	// Version of LMMs that created this project
+	ProjectVersion creator =
+		documentElement().attribute( "creatorversion" ).
+		replace( "svn", "" );
+
+	// Vector of all versions that have upgrade routines.
+	std::vector<QString> upgradeVersions = {
+		"0.2.1-20070501"   ,   "0.2.1-20070508"   ,   "0.3.0-rc2",
+		"0.3.0"            ,   "0.4.0-20080104"   ,   "0.4.0-20080118",
+		"0.4.0-20080129"   ,   "0.4.0-20080409"   ,   "0.4.0-20080607",
+		"0.4.0-20080622"   ,   "0.4.0-beta1"      ,   "0.4.0-rc2",
+		"1.0.99-0"         ,   "1.1.0-0"          ,   "1.1.91-0",
+		"1.2.0-rc3"        ,   "1.3.0"
+	};
+
+	// Get an iterator pointing at the first upgrade we need to run (or at upgrades.end() if there is no such upgrade)
+	auto firstRequiredUpgrade = std::find_if( upgradeVersions.begin(), upgradeVersions.end(),
+		[creator](QString version){ return creator < version; }
+	);
```

Comment:
Nice use of standard library algorithms. However, there's a specific one for this purpose: [`std::upper_bound`](https://en.cppreference.com/w/cpp/algorithm/upper_bound).